File: 5. LinkedLists/sum_list/sum_list.js
// class Node {
//   constructor(val) {
//     this.val = val;
//     this.next = null;
//   }
// }

// const a = new Node(2);
// const b = new Node(8);
// const c = new Node(3);
// const d = new Node(-1);
// const e = new Node(7);

// a.next = b;
// b.next = c;
// c.next = d;
// d.next = e;



// Solutions
// ---------
// Iterative
const sumList = (head) => {
    let sum = 0;
    while(head !== null){
        sum += head.val;
        head = head.next;
    }
    return sum;
}


// Recursive
// const sumList = (head) => {
//     if(head === null) return 0;
//     return head.val + sumList(head.next);
// }


// let ans = sumList(a);
// console.log(`Ans: ${ans}`);

// Complexity:
// Iterative:
// n = number of nodes
// Time: O(n)
// Space: O(1)

// Recursive:
// n = number of nodes
// Time: O(n)
// Space: O(n)

File: 5. LinkedLists/2. sum_list/sum_list.py
class Node:
    def __init__(self, value):
        self.val = value
        self.next = None

a = Node(2)
b = Node(8)
c = Node(3)
d = Node(-1)
e = Node(7)

a.next = b
b.next = c
c.next = d
d.next = e

# 2 -> 8 -> 3 -> -1 -> 7

def print_list(head):
  output = ""
  while(head.next is not None):
    output += str(head.val) + "->"
    head = head.next
  output += str(head.val)
  print(output)

# Iterative
def sum_list(head):
    current = head
    sum = 0
    while(current is not None):
        sum += current.val
        current = current.next
    return sum

# Recursive
def sum_list1(head):
    if head is None:
        return 0
    return head.val + sum_list1(head.next)

print_list(a)
print(sum_list(a))
print(sum_list1(a))